src/Shell/Task/MigrationDiffTask.php

```diff
@@ -0,0 +1,490 @@
+     * state of the database and the schema dump file.
+     *
+     * @return void
+     */
+    protected function calculateDiff()
+    {
+        $this->getConstraints();
+        $this->getIndexes();
+        $this->getColumns();
+        $this->getTables();
+    }
+
+    /**
+     * Calculate the diff between the current state of the database and the schema dump
+     * by returning an array containing the full \Cake\Database\Schema\Table definitions
+     * of tables to be created and removed in the diff file.
+     *
+     * The method directly sets the diff in a property of the class.
+     *
+     * @return void
+     */
+    protected function getTables()
+    {
+        $this->templateData['fullTables'] = [
+            'add' => array_diff_key($this->currentSchema, $this->dumpSchema),
+            'remove' => array_diff_key($this->dumpSchema, $this->currentSchema)
+        ];
+    }
+
+    /**
+     * Calculate the diff between columns in existing tables.
+     * This will look for columns addition, columns removal and changes in columns metadata
+     * such as change of types or property such as length.
+     *
+     * Note that the method is not able to detect columns name change.
+     * The method directly sets the diff in a property of the class.
+     *
+     * @return void
+     */
+    protected function getColumns()
+    {
+        foreach ($this->commonTables as $table => $currentSchema) {
+            $currentColumns = $currentSchema->columns();
+            $oldColumns = $this->dumpSchema[$table]->columns();
+
+            // brand new columns
+            $addedColumns = array_diff($currentColumns, $oldColumns);
+            foreach ($addedColumns as $columnName) {
+                $this->templateData[$table]['columns']['add'][$columnName] = $currentSchema->column($columnName);
+            }
+
+            // changes in columns meta-data
+            foreach ($currentColumns as $columnName) {
+                $column = $currentSchema->column($columnName);
+                $oldColumn = $this->dumpSchema[$table]->column($columnName);
+
+                if (in_array($columnName, $oldColumns) &&
+                    $column !== $oldColumn
+                ) {
+                    $changedAttributes = array_diff($column, $oldColumn);
+
+                    if (!isset($changedAttributes['type'])) {
+                        $changedAttributes['type'] = $column['type'];
+                    }
+                    $this->templateData[$table]['columns']['changed'][$columnName] = $changedAttributes;
+                }
+            }
+
+            // columns deletion
+            if (!isset($this->templateData[$table]['columns']['remove'])) {
+                $this->templateData[$table]['columns']['remove'] = [];
+            }
+            $removedColumns = array_diff($oldColumns, $currentColumns);
+            if (!empty($removedColumns)) {
```

Comment:
`if ($removedColumns) {` as per new guideline to avoid spelling mistakes
but no big deal either way.